### PR TITLE
Extract rendering strategy for decorations

### PR DESCRIPTION
**Pull request by AlanGriffiths**
This extracts the generic rendering integration from the default "strategy" for rendering

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
Note that while this is a stable intermediate form (and can be merged), it doesn't extract everything that should be customised by a "decoration strategy": There's still two customisation points needed: the "static geometry", and the button placement

**Review by mattkae — CHANGES_REQUESTED** [review]
Some initial feedback but it looks like a good move to me!

File: src/server/shell/decoration/renderer_default.cpp [hunk 1]
```diff
@@ -0,0 +1,820 @@
...
+    char const* const font_path_search_paths[]{
+        "/usr/share/fonts/truetype",    // Ubuntu/Debian
+        "/usr/share/fonts/TTF",         // Arch
+        "/usr/share/fonts",             // Fedora/Arch
+    };
+
+    std::vector<std::string> usable_search_paths;
+    for (auto const& path : font_path_search_paths)
+    {
+        if (boost::filesystem::exists(path))
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/server/shell/decoration/renderer_default.cpp:100]: This is existing code I didn't aim to rework for this PR

File: src/server/shell/decoration/renderer_default.cpp [hunk 2]
```diff
@@ -0,0 +1,820 @@
...
+    }
+
+    for (auto const& font : font_paths)
+    {
+        for (auto const& prefix : font.prefixes)
+        {
+            for (auto const& path : usable_search_paths)
+            {
+                auto const full_font_path = path + '/' + prefix + '/' + font.filename;
+                if (boost::filesystem::exists(full_font_path))
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/server/shell/decoration/renderer_default.cpp:111]: This is existing code I didn't aim to rework for this PR

File: src/server/shell/decoration/renderer_default.cpp [hunk 3]
```diff
@@ -0,0 +1,820 @@
...
+    }
+
+    return "";
+}
+
+inline auto area(geom::Size size) -> size_t
+{
+    return (size.width > geom::Width{} && size.height > geom::Height{})
+        ? size.width.as_int() * size.height.as_int()
+        : 0;
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/server/shell/decoration/renderer_default.cpp:124]: This is existing code I didn't aim to rework for this PR

File: src/server/shell/decoration/renderer_default.cpp [hunk 4]
```diff
@@ -0,0 +1,820 @@
...
+    Pixel color)
+{
+    if (!area(buf_size) || height_pixels <= geom::Height{})
+        return;
+
+    std::lock_guard lock{mutex};
+
+    if (!library || !face)
+    {
+        mir::log_warning("FreeType not initialized");
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/server/shell/decoration/renderer_default.cpp:416]: This is existing code I didn't aim to rework for this PR

File: src/server/shell/decoration/renderer_default.cpp [hunk 5]
```diff
@@ -0,0 +1,820 @@
...
+        return;
+    }
+
+    try
+    {
+        set_char_size(height_pixels);
+    }
+    catch (std::runtime_error const& error)
+    {
+        mir::log_warning("%s", error.what());
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/server/shell/decoration/renderer_default.cpp:426]: This is existing code I didn't aim to rework for this PR

File: src/server/shell/decoration/renderer_default.cpp [hunk 6]
```diff
@@ -0,0 +1,820 @@
...
+    void redraw_titlebar_text(geom::Size scaled_titlebar_size);
+    void redraw_titlebar_buttons(geom::Size scaled_titlebar_size);
+};
+}
+
+auto msd::RendererStrategy::default_strategy(std::shared_ptr<StaticGeometry const> const& static_geometry)
+-> std::unique_ptr<RendererStrategy>
+{
+    return std::make_unique<::RendererStrategy>(static_geometry);
+}
```
[the remark this answers is not in the document]
Comment (AlanGriffiths) [reply src/server/shell/decoration/renderer_default.cpp:297]: This just wires things up. It goes away in the follow-up PR

**Review by mattkae — APPROVED** [review]
Approving since this is all preexisting :)

**Pull request merged**